samples/order/processor/processor.go

```diff
@@ -0,0 +1,251 @@
+	}
+	if err := sendJSON(ctx, p.responseTopic, res); err != nil {
+		// TODO(jba): should we really ack in this case? Not clear.
+		return err
+	}
+	log.Printf("sent %+v", res)
+	return nil
+}
+
+// handleOrder processes the order request. A processing error is a kind of response.
+func (p *processor) handleOrder(ctx context.Context, req *common.OrderRequest) *common.OrderResponse {
+	res, err := p.processOrder(ctx, req)
+	if err != nil {
+		// TODO(jba): record error metric
+		res = &common.OrderResponse{
+			ID:   req.ID,
+			Note: fmt.Sprintf("processing failed: %v", err),
+		}
+	}
+	return res
+}
+
+func (p *processor) processOrder(ctx context.Context, req *common.OrderRequest) (res *common.OrderResponse, err error) {
+	// See if there is already a document for this order.
+	order := &common.Order{ID: req.ID}
+	err = p.coll.Get(ctx, order)
+	switch {
+	case gcerrors.Code(err) == gcerrors.NotFound:
+		// Normal case: the order hasn't been created yet. Do so.
+		order = &common.Order{
+			ID:         req.ID,
+			Email:      req.Email,
+			InImage:    req.InImage,
+			OutImage:   fmt.Sprintf("%s-out.png", req.InImage),
+			CreateTime: req.CreateTime,
+		}
+		if err := p.coll.Create(ctx, order); err != nil {
+			return nil, err
+		}
+
+	case err == nil && order.FinishTime.IsZero():
+		// The order exists, but was not finished. Either it was abandoned by the processor that
+		// was working on it (probably because the processor died), or it is in progress. Assume
+		// that it was abandoned, and process it.
+		// There is nothing to do here, since all the existing order fields are valid.
+
+	case err == nil && !order.FinishTime.IsZero():
+		// The order exists and was finished. This is most likely the result of a pubsub redelivery.
+		// We simply ignore it.
+		return nil, nil
+	}
+	// At this point, there is an unfinished Order with ID == req.ID in the database.
+
+	defer func() {
+		// Mark the order complete by updating the finish time.
+		err2 := p.coll.Update(ctx, order, docstore.Mods{"FinishTime": time.Now()})
+		if err2 != nil {
+			if err == nil {
+				err = err2
+			} else {
+				err = fmt.Errorf("%v and %v", err, err2)
+			}
+		}
+	}()
+
+	// Read the input image from the bucket.
+	r, err := p.bucket.NewReader(ctx, req.InImage, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	defer r.Close()
+	img, format, err := image.Decode(r)
+	if err != nil {
+		return nil, err
+	}
+
+	// Process and write the output image.
+	w, err := p.bucket.NewWriter(ctx, order.OutImage, nil)
+	if err != nil {
+		return nil, err
+	}
+	if err := png.Encode(w, img); err != nil {
+		w.Close()
+		return nil, err
+	}
+	if err := w.Close(); err != nil {
+		return nil, err
+	}
+
+	return &common.OrderResponse{
+		ID:       req.ID,
+		OutImage: order.OutImage,
+		Note:     fmt.Sprintf("converted from %s to png", format),
+	}, nil
+}
+
+// Marshal msg to JSON and send it over topic.
+func sendJSON(ctx context.Context, topic *pubsub.Topic, msg interface{}) error {
```

Comment:
Nit: I'd consider inlining these, they are pretty short and they hide the punchline of sending/receiving to the topic/subscription a bit.

Reply:
done

Reply:
https://github.com/google/go-cloud/pull/2444/files#diff-39561593e59e6187e8f7eea4a0db8dffR58